File: 130-surrounded-regions/surrounded-regions.cpp
class Solution
{
    public:
        void dfs(int i, int j, vector<vector < char>> &board, vector< vector< int>> &vis)
        {
            if (i < 0 or j < 0 or i >= board.size() or
             j >= board[0].size() or board[i][j] == 'X' or vis[i][j]==1)
                return;

            vis[i][j] = 1;

            dfs(i + 1, j, board, vis);
            dfs(i - 1, j, board, vis);
            dfs(i, j - 1, board, vis);
            dfs(i, j + 1, board, vis);
        }
    void solve(vector<vector < char>> &board)
    {
        int m = board.size();
        int n = board[0].size();

        vector<vector < int>> vis(m, vector<int> (n, 0));

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if ((i == 0 or i == m - 1 or j == 0 or j == n - 1) and vis[i][j] == 0 and board[i][j] == 'O')
                {
                    dfs(i, j, board, vis);
                }
            }
        }

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (vis[i][j] == 0 and board[i][j] == 'O')
                    board[i][j] = 'X';
            }
        }
    }
};

File: 130-surrounded-regions/surrounded-regions.java
class Solution {
    public void dfs(int row , int col ,  boolean visited[][] ,char[][] board, int n, int m )
    {
        visited[row][col] = true;
        int delRow[] = {1,0,-1,0};
        int delCol[] = {0,-1,0,1};

        for(int i=0;i<4;i++)
        {
            int nRow = row + delRow[i];
            int nCol = col + delCol[i];

            if(nRow>=0 && nCol>=0 && nRow<n && nCol<m && visited[nRow][nCol] == false && board[nRow][nCol]=='O')
            {
                dfs(nRow,nCol,visited,board,n,m);
            }
        }
    }
    public void solve(char[][] board) {
        int n = board.length;
        int m = board[0].length;
        boolean visited[][] = new boolean[n][m];

        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(i==0 || j==0 || i==n-1 || j == m-1 )
                {
                    if(!visited[i][j] && board[i][j] == 'O')
                    dfs(i,j,visited,board,n,m);
                }
            }
        }
        
         for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(!visited[i][j] && board[i][j] == 'O')
                {
                    board[i][j] = 'X';
                }
            }
        }



    }
}